### PR TITLE
[LMBiSeNet] Implement LMBiSeNet

### PR DESCRIPTION
**Pull request by ruimashita**
Implement new semantic segmentation model `LMBiSeNet` that is 5 times faster and little better `mean IoU` than `LMSegNetV1`.

model's proto buffer: https://drive.google.com/open?id=17VMylt0INkaPfrT9nxbCw2hrTs7Hy0Ln

Summary of experiment results 
* Image size: {height: 352, width: 480}
* Number of class: 11

Quantized model size:

| Param | Size (MB) | 1 bits Quant Size (MB) | FLOPs (m) |
| --: | --: | --: | --: |
| **6914497** | **26.37671** | **0.91155** | **12518.1925** |

Quantized model accuracy on Camvid validation:

|  **dataset** | **global accuracy** | **mean accuracy** | **mean iou** |
| ----: | ----: | ----: | ----: |
|  **Camvid** | 0.895 | 0.708 | 0.590 |

## Motivation and Context

The inference time of Quantized LMSegNetV1 on DE10-Nano takes about 3.5 second as 352x480 input image size, in spite of low accuracy.
To accelerate this time, I tried to create new semantic segmentation model.

## Description

* Implement `LMBiSeNet` class.
* Implement `Softmax` post processor
* Implement unit test for verifing to just train `LMBiSeNet` with no error, and post process on training (using tf method) is same as on inference (using post processor).

The new network architecture reference to mainly [BiSeNet](https://arxiv.org/abs/1808.00897), [Fast-SCNN](https://arxiv.org/abs/1902.04502) and [ContextNet](https://arxiv.org/abs/1805.04554) also.

Major difference from BiSeNet:
    * Apply first convolution then branch into contextual and spatial path.
    * Use space to depth (s2d) and depth to space (d2s) as upsample and downsample.
    * Use only 1 stride 1x1 and 3x3 convolution (not use multi stride and dilated convolution)
        for be limited by our convolution IP.
    * Use DensNet block in contextual part.
    * All of convolution out channels are less than BiSeNet for inference time.
    * Use attention refinement module reference to BiSeNet after last layer of 1/32
        (BiSeNet: both 1/16 and 1/32) in context path.
    * Use relu activation instead of sigmoid in attention refinement and feature fusion module.
    * In up-sampling followed by context path, alternate d2s and 1x1 conv for reducing channel size.

### 1. Training results with this command.
`$ docker-compose -p lmnet run --rm tensorflow python3 executor/train.py -c configs/core/segmentation/lm_bisenet_quantize_camvid.py`

Screen Shot 2019-04-10 at 22 01 13

### 2. Run `lm_x86.elf` on x86 server.

```
$  ./models/lib/lm_x86.elf ../inference_test_data/000_images_placeholder:0.npy ../inference_test_data/731_output:0.npy
-------------------------------------------------------------
Comparison: Default network test  failed...
Failed count: 1
First failed report
index: 956 / 29040
input: -22.6264, expected: -22.6264

-------------------------------------------------------------
Add,114,  sum:0.114ms
ApplyThresholds,3717,1687,681,1554,605,1391,610,1397,700,224,583,156,760,152,1775,1323,639,285,277,255,301,253,294,623,1240,7828,5789,2925,  sum:38.024ms
AveragePool,444,440,468,  sum:1.352ms
BatchNorm,6792,169,10,10,201,1,1,  sum:7.184ms
Convolution,37422,9,7,2141,  sum:39.579ms
ExtractImagePatches,8363,479,176,124,128,77,4639,1245,742,  sum:15.973ms
MulDepthWise,139,136,184,  sum:0.459ms
QTZ_linear_mid_tread_half,15306,84,433,  sum:15.823ms
QuantizedConv2D,14459,492,5574,  sum:20.525ms
QuantizedConv2D_ApplyScalingFactor,234,2,488,  sum:0.724ms
ReLu,107,1,103,0,0,  sum:0.211ms
TotalInitTime,34190,  sum:34.19ms
TotalRunTime,650778,  sum:650.778ms
func_ConcatOnDepth,672,724,881,289,262,312,151,217,249,448,932,  sum:5.137ms
kn2row-1x1,37420,8,7,2141,  sum:39.576ms
matrix_multiplication,37417,7,6,2140,  sum:39.57ms
matrix_shift_add1,174,151,154,133,65,71,194,209,194,623,658,501,  sum:3.127ms
matrix_shift_add2,513,418,438,201,93,100,87,90,83,5883,2981,955,  sum:11.842ms
pack_input_to_qwords,10964,565,253,598,240,677,218,798,229,114,257,96,240,105,287,218,232,105,46,150,46,200,46,231,1,184,92,6098,1481,1105,833,  sum:26.709ms
quantized-kn2row,73338,4649,10166,5326,8029,6875,6595,13503,2929,3711,3468,3696,3388,3400,6802,2894,9418,2262,8196,3844,7550,3779,10506,14443,490,3884,2133,115314,101021,115130,5557,  sum:562.296ms
quantized_matrix_multiplication,58593,2388,7319,3158,5879,4793,5050,11292,1990,2967,2612,3251,2379,2934,4730,1347,8508,1859,7469,3424,6805,3305,9796,14138,413,3069,797,86078,83581,108092,4718,  sum:462.734ms
quantized_ohwi_to_hwoi,16,1,10,8,2,8,2,8,2,10,9,2,2,2,3,1,29,4,23,6,29,10,23,53,49,1,0,2,5,20,1,  sum:0.341ms
```

I don't know why `input: -22.6264, expected: -22.6264` is fail, It looks same values...

### 3. Run inference `lm_arm.elf` on FPGA.
```
root@DE10_NANO:~/lm_bisenet_camvid/export/save.ckpt-18350/352x480/output# ./models/lib/lm_arm.elf ../inference_test_data/000_images_placeholder\:0.npy ../inference_test_data/731_output\:0.npy
-------------------------------------------------------------
Comparison: Default network test  succeeded!!!
-------------------------------------------------------------
Add,2456,  sum:2.456ms
ApplyThresholds,33785,12118,4995,11865,4953,11917,5018,11784,5087,920,5138,964,5089,902,12454,11644,5165,2479,2469,2411,2441,2351,2523,5262,12126,167113,94081,27502,  sum:464.556ms
AveragePool,7445,7484,6670,  sum:21.599ms
BatchNorm,55560,2219,241,194,1417,64,25,  sum:59.72ms
Convolution,61928,87,80,15918,  sum:78.013ms
ExtractImagePatches,20080,2071,565,669,552,322,26554,6735,5126,  sum:62.674ms
MulDepthWise,2013,2034,2139,  sum:6.186ms
Pack_input_for_tiling,147611,8921,4400,11221,4430,13349,4419,15664,4461,2201,4986,2189,5470,2197,6146,4435,4874,2379,1195,3526,1215,4753,1218,6011,82,4496,2277,145342,35970,27036,20108,  sum:502.582ms
QTZ_linear_mid_tread_half,52066,430,2007,  sum:54.503ms
QuantizedConv2D,53591,1954,37394,  sum:92.939ms
QuantizedConv2D_ApplyScalingFactor,4326,30,4350,  sum:8.706ms
ReLu,1795,14,1772,4,5,  sum:3.59ms
TotalInitTime,312616,  sum:312.616ms
TotalRunTime,2.25565e+06,  sum:2255.65ms
func_ConcatOnDepth,5211,6154,7051,2130,2362,2586,1341,1781,2233,4470,9126,  sum:44.445ms
kn2row-1x1,61917,78,73,15912,  sum:77.98ms
matrix_multiplication,61905,69,67,15900,  sum:77.941ms
matrix_transpose (row_major),58,  sum:0.058ms
```

### 4. Run on FPGA.
```
$ root@DE10_NANO:~/lm_bisenet_camvid/export/save.ckpt-18350/352x480/output# ./models/lib/lm_fpga.elf ../inference_test_data/000_images_placeholder\:0.npy ../inference_test_data/731_output\:0.npy
-------------------------------------------------------------
Comparison: Default network test  failed...
Failed count: 29040
First failed report
index: 0 / 29040
input: -7.08966, expected: -7.80849

-------------------------------------------------------------
Add,2436,  sum:2.436ms
AveragePool,7456,7367,6627,  sum:21.45ms
BatchNorm,44120,2242,233,261,1394,68,26,  sum:48.344ms
Convolution,61820,86,81,15997,  sum:77.984ms
ExtractImagePatches,20091,2061,557,587,561,326,26626,6741,5247,  sum:62.797ms
MulDepthWise,2249,2175,2202,  sum:6.626ms
Packing input for kn2row,21586,1349,596,1762,598,2042,595,2382,595,246,656,196,710,240,880,594,632,267,112,290,119,519,97,790,16,597,288,21544,5509,4128,3142,  sum:73.077ms
QConv2D kn2row tiling,9363,1028,2008,1361,1961,1260,1963,3126,902,842,1059,803,1111,798,1907,959,3788,1273,4335,1633,4334,1994,4336,8108,6779,1007,965,29860,21581,17550,1584,  sum:139.578ms
QTZ_linear_mid_tread_half,52166,354,2005,  sum:54.525ms
QuantizedConv2D,10901,6917,7025,  sum:24.843ms
QuantizedConv2D_ApplyScalingFactor,4113,35,3701,  sum:7.849ms
ReLu,1798,14,1772,4,5,  sum:3.593ms
TotalInitTime,278156,  sum:278.156ms
TotalRunTime,646152,  sum:646.152ms
func_ConcatOnDepth,5201,6166,7087,2178,2334,2568,1343,1786,2237,4452,9140,  sum:44.492ms
kn2row-1x1,61808,78,75,15990,  sum:77.951ms
matrix_multiplication,61795,69,68,15977,  sum:77.909ms
matrix_transpose (row_major),96,  sum:0.096ms
pack_input_to_qwords,21569,1332,582,1750,581,2030,583,2371,577,236,647,186,700,231,870,585,577,257,104,281,110,511,89,781,9,587,279,21530,5495,4116,3128,  sum:72.684ms
```

The results are significant difference from expected. Need to fix.

### 5. Check model size
```
$ docker-compose -p lmnet run --rm cpu-tensorflow python3 executor/profile_model.py -c configs/core/segmentation/lm_bisenet_quantize_camvid.py  -b 1 -uql block_first -uql block_last -uql fusion/float_conv_1 -uql fusion/float_conv_2
```

* Image size: {height: 352, width: 480}
* Number of class: 11

| Name | Param | Size (MB) | 1 bits Quant Size (MB) | FLOPs (m) |
| :-- | --: | --: | --: | --: |
| **total** | **6914497** | **26.37671** | **0.91155** | **12518.1925** |

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: lmnet/configs/core/segmentation/lm_bisenet_quantize_camvid.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 The Blueoil Authors. All Rights Reserved.
```
Comment (nlpng) [inline lmnet/configs/core/segmentation/lm_bisenet_quantize_camvid.py]: Not certain about this, but should this be `Copyright 2019` or not?

Comment (ruimashita) [reply lmnet/configs/core/segmentation/lm_bisenet_quantize_camvid.py]: thanks 👍

Comment (ruimashita) [reply lmnet/configs/core/segmentation/lm_bisenet_quantize_camvid.py]: fixed in https://github.com/blue-oil/blueoil/pull/225/commits/d3526e900ba746e8c678b91ac4a6fee927152f2b

**Comment by ruimashita** [comment]
@iizukak 
Thanks :smile: :smile: Sounds nice :100: 
I will post a issue.

**Comment by iizukak** [comment]
@nlpng Please review again 👀
